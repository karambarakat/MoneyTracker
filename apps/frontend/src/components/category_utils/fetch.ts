import { UseQueryResult, useQuery } from '@tanstack/react-query'
import { useMemo } from 'react'

export const iconQuery = async () => {
  const res = await import('./icons')

  return res.default.split('\n') as string[]
}

/**
 *
 * @param whichOne select specific icon or return all icons
 */
export function useIcon(whichOne: string) {
  const res = useQuery({
    queryFn: iconQuery,
    suspense: false,
    queryKey: ['icons'],
    cacheTime: 1000 * 60 * 60 * 24,
  }) as UseQueryResult<string>
  // return data of type string please

  const validIndex = useMemo(() => {
    const target = Number(whichOne)

    if (isNaN(target) || target > 876 || target < 0)
      return new Error('Invalid index')

    return target
  }, [whichOne])

  if (validIndex instanceof Error) return validIndex

  if (res.status === 'success')
    return { status: 'success', data: res.data[validIndex] }

  return res
}

export const defaultValue =
  'm38.49,32h-14.49V8h24v18h-6.04l-3.46,6Zm-18.49,4v-14c-7.73,0-14,6.27-14,14s6.27,14,14,14,14-6.27,14-14h-14Zm27.73-6h-3.46l-12.12,21c.77,1.33.96,1.67,1.73,3h24.25c.77-1.33.96-1.67,1.73-3l-12.12-21Z'
