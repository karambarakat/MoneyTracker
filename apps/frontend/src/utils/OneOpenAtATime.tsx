import { useId } from '@mantine/hooks'
import { createContext, useContext, useEffect, useState } from 'react'
import { WithChildren } from 'ui/src/utils/WithChildren'

const context = createContext<{
  open: false | string
  setOpen: (s: string | false) => void
}>({ open: false, setOpen: () => console.error('no context assigned') })

export const useOneState = (initial?: boolean) => {
  const id = useId()
  const { open, setOpen } = useContext(context)
  const [_open, _setOpen] = useState(open === id)
  useEffect(() => {
    if (initial) setOpen(id)
  }, [initial])
  useEffect(() => {
    if (open === id) _setOpen(true)
    else _setOpen(false)
  }, [open])

  return [
    _open,
    function (arg: boolean | ((old: boolean) => boolean)) {
      if (typeof arg === 'function') {
        const rt = arg(_open)
        return rt ? setOpen(id) : setOpen(false)
      }
      if (arg) setOpen(id)
      else setOpen(false)
    },
  ] as const
}

export const OneStateProvider = ({ children }: WithChildren) => {
  const [open, setOpen] = useState<false | string>(false)
  return (
    <context.Provider value={{ open, setOpen }}>{children}</context.Provider>
  )
}
