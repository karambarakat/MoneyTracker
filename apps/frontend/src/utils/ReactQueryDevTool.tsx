import { ReactQueryDevtools } from '@tanstack/react-query-devtools'
import React from 'react'

const ReactQueryDevtoolsProductionLazy = React.lazy(() =>
  import('@tanstack/react-query-devtools/build/lib/index.prod.js').then(d => ({
    default: d.ReactQueryDevtools,
  })),
)

export function ReactQueryDevtoolsProduction() {
  const [showDevtools, setShowDevtools] = React.useState(false)

  React.useEffect(() => {
    // @ts-ignore
    window.toggleDevtools = () => setShowDevtools(old => !old)
  }, [])

  return showDevtools ? (
    <React.Suspense fallback={null}>
      <ReactQueryDevtoolsProductionLazy />
    </React.Suspense>
  ) : null
}

export default function ReactQueryDevTools_() {
  return (
    <>
      <ReactQueryDevtools initialIsOpen={false} position="bottom-right" />
      <ReactQueryDevtoolsProduction />
    </>
  )
}
