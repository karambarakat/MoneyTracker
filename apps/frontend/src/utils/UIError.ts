export default class UIError extends Error {
  jsError: unknown
  constructor(message: string, jsError?: unknown) {
    super(message)
    this.jsError = jsError
  }
}
