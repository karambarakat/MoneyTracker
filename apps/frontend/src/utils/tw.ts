// this is temprorary until I switch to UnoCSS or
// figure out a way to simplify tw configuration completely

import tw from 'twin.macro'

export const iconRed = tw`fill-red-800 dark:fill-red-400`

export const iconBlue = tw`fill-blue-800 dark:fill-blue-400`
