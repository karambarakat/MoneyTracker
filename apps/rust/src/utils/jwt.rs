#[allow(dead_code)]
// for some limitation for ts_rs, this exists for the sake of simplicity
#[derive(ts_rs::TS)]
#[ts(export)]
pub struct Jwt_ {
    pub id: String,
    pub email: String,
    pub user_id: String,
    pub exp: String,
    pub iat: String,
}

pub struct Jwt {
    pub id: String,
    pub email: String,
    pub user_id: String,
    pub exp: DateTime<Utc>,
    pub iat: DateTime<Utc>,
}
use chrono::DateTime;
use chrono::Duration;
use chrono::Utc;
use hmac::{Hmac, Mac};
use jwt::SignWithKey;
use jwt::VerifyWithKey;
use sha2::Sha256;
use std::collections::BTreeMap;

use crate::errors::MyErrors;

impl Jwt {
    pub fn validate(str: &str) -> Result<Self, MyErrors> {
        let env = std::env::var("JWT_SALT").expect("JWT_SALT is not set");

        let key: Hmac<Sha256> = Hmac::new_from_slice(env.as_bytes())
            .map_err(|err| MyErrors::InternalError(Box::new(err)))?;

        let claims: BTreeMap<String, String> = str
            .verify_with_key(&key)
            .map_err(|err| MyErrors::InternalError(Box::new(err)))?;

        let exp = claims.get("exp");

        let exp = match exp {
            Some(exp) => {
                // test if expirted
                let exp = exp
                    .parse::<i64>()
                    .map_err(|err| MyErrors::InternalError(Box::new(err)))?;
                let exp = DateTime::<Utc>::from_utc(
                    chrono::NaiveDateTime::from_timestamp_opt(exp, 0).unwrap(),
                    Utc,
                );

                if exp < Utc::now() {
                    return Err(MyErrors::ExpiredBearerToken);
                }

                exp
            }
            None => return Err(MyErrors::ExpiredBearerToken),
        };

        Ok(Self {
            id: claims.get("id").unwrap().to_string(),
            email: claims.get("email").unwrap().to_string(),
            user_id: claims.get("user_id").unwrap().to_string(),
            exp,
            iat: DateTime::<Utc>::from_utc(
                chrono::NaiveDateTime::from_timestamp_opt(
                    claims
                        .get("iat")
                        .unwrap()
                        .parse::<i64>()
                        .map_err(|err| MyErrors::InternalError(Box::new(err)))?,
                    0,
                )
                .unwrap(),
                Utc,
            ),
        })
    }

    pub fn sign(id: &str, email: &str, user_id: &str) -> Result<String, MyErrors> {
        let env = std::env::var("JWT_SALT").expect("JWT_SALT is not set");

        let key: Hmac<Sha256> = Hmac::new_from_slice(env.as_bytes())
            .map_err(|err| MyErrors::InternalError(Box::new(err)))?;

        let claims = {
            let mut claims: BTreeMap<String, String> = BTreeMap::new();

            let now = Utc::now();
            let exp = now + Duration::days(1);

            claims.insert("id".to_string(), id.to_string());
            claims.insert("email".to_string(), email.to_string());
            claims.insert("user_id".to_string(), user_id.to_string());
            claims.insert("exp".to_string(), exp.timestamp().to_string());
            claims.insert("iat".to_string(), now.timestamp().to_string());

            claims
        };

        let token_str = claims
            .sign_with_key(&key)
            .map_err(|err| MyErrors::InternalError(Box::new(err)))?;

        Ok(token_str)
    }
}

impl std::fmt::Display for Jwt {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "[User: {},****]", self.email)
    }
}
