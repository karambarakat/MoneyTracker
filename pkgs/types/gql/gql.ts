/* eslint-disable */
import * as types from './graphql';
import { TypedDocumentNode as DocumentNode } from '@graphql-typed-document-node/core';

/**
 * Map of all GraphQL operations in the project.
 *
 * This map has several performance disadvantages:
 * 1. It is not tree-shakeable, so it will include all operations in the project.
 * 2. It is not minifiable, so the string of a GraphQL query will be multiple times inside the bundle.
 * 3. It does not support dead code elimination, so it will add unused operations.
 *
 * Therefore it is highly recommended to use the babel or swc plugin for production.
 */
const documents = {
    "\n  fragment Category on Category {\n    id\n    title\n    color\n    icon\n  }\n": types.CategoryFragmentDoc,
    "\n  fragment User on User {\n    avatar\n    createdAt\n    displayName\n    email\n    id\n    providers\n    updatedAt\n  }\n": types.UserFragmentDoc,
    "\n  fragment Entry on Entry {\n    id\n    title\n    amount\n    note\n    createdAt\n    updatedAt\n    category {\n      id\n      title\n      color\n      icon\n    }\n  }\n": types.EntryFragmentDoc,
};

/**
 * The graphql function is used to parse GraphQL queries into a document that can be used by GraphQL clients.
 *
 *
 * @example
 * ```ts
 * const query = graphql(`query GetUser($id: ID!) { user(id: $id) { name } }`);
 * ```
 *
 * The query argument is unknown!
 * Please regenerate the types.
 */
export function graphql(source: string): unknown;

/**
 * The graphql function is used to parse GraphQL queries into a document that can be used by GraphQL clients.
 */
export function graphql(source: "\n  fragment Category on Category {\n    id\n    title\n    color\n    icon\n  }\n"): (typeof documents)["\n  fragment Category on Category {\n    id\n    title\n    color\n    icon\n  }\n"];
/**
 * The graphql function is used to parse GraphQL queries into a document that can be used by GraphQL clients.
 */
export function graphql(source: "\n  fragment User on User {\n    avatar\n    createdAt\n    displayName\n    email\n    id\n    providers\n    updatedAt\n  }\n"): (typeof documents)["\n  fragment User on User {\n    avatar\n    createdAt\n    displayName\n    email\n    id\n    providers\n    updatedAt\n  }\n"];
/**
 * The graphql function is used to parse GraphQL queries into a document that can be used by GraphQL clients.
 */
export function graphql(source: "\n  fragment Entry on Entry {\n    id\n    title\n    amount\n    note\n    createdAt\n    updatedAt\n    category {\n      id\n      title\n      color\n      icon\n    }\n  }\n"): (typeof documents)["\n  fragment Entry on Entry {\n    id\n    title\n    amount\n    note\n    createdAt\n    updatedAt\n    category {\n      id\n      title\n      color\n      icon\n    }\n  }\n"];

export function graphql(source: string) {
  return (documents as any)[source] ?? {};
}

export type DocumentType<TDocumentNode extends DocumentNode<any, any>> = TDocumentNode extends DocumentNode<  infer TType,  any>  ? TType  : never;