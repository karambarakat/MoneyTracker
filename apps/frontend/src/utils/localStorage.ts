import { useEffect, useState } from 'react'
import { Jwt_ } from 'types/backend'
import { User } from 'types/gql/graphql'

const eventTarget = new EventTarget()

class LocalItem<T> {
  /**
   * @private
   */
  _item: T | undefined
  key: string
  constructor(key: string) {
    this.key = key
    const localItem = localStorage.getItem(key)
    this._item = localItem ? JSON.parse(localItem as string) : undefined
  }

  setItem(item: T | undefined) {
    if (!item) localStorage.removeItem(this.key)
    else localStorage.setItem(this.key, JSON.stringify(item))

    this._item = item

    // React
    eventTarget.dispatchEvent(new Event(this.key))
  }

  getItem() {
    const item = localStorage.getItem(this.key)

    if (!item) return

    return JSON.parse(item) as T
  }

  use() {
    const [item, setItem] = useState<LocalItem<T>>(this)

    useEffect(() => {
      const handler = () => setItem(new LocalItem(this.key))
      eventTarget.addEventListener(this.key, handler)
      return () => eventTarget.removeEventListener(this.key, handler)
    }, [])

    return item
  }
}

export const profile = new LocalItem<User>('profile')

export const token = Object.assign(new LocalItem<string>('token'), {
  valid(this: LocalItem<string>) {
    const item = this.getItem()

    if (!item) return false

    return (
      Number((JSON.parse(atob(item.split('.')[1])) as Jwt_).exp) >=
      Date.now() / 1000
    )
  },
})
