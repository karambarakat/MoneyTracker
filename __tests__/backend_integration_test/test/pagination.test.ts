import { afterAll, beforeAll, expect, it } from '@jest/globals'
import { fetchQql, gql } from '../utils/gql'
import fetch from 'node-fetch'
import { HELPER_SERVER, REST_API } from '../utils/constants'
import { categoryGQL, entryGQL, entryInput, entryNoCat } from '../utils/expect'
import {
  Mutation,
  MutationCreateManyEntriesArgs,
  Query,
  QueryGetAllEntriesArgs,
} from 'types/gql/graphql'
import { inspect } from 'util'

const ctx: any = {
  ids: [],
  credential: 'testing_pagination@m.c:pass',
}

beforeAll(async () => {
  const res = await fetch(REST_API + '/auth/local/register', {
    method: 'POST',
    headers: {
      content: 'application/json',
      authorization: 'Basic ' + Buffer.from(ctx.credential).toString('base64'),
    },
    body: JSON.stringify({}),
  })
  const json: any = await res.json()
  ctx.user = Object.assign({}, json)
  ctx.ids.push(json.id)
  ctx.user.token = res.headers.get('x-token')
  ctx.user.headers = {
    authorization: 'Bearer ' + ctx.user.token,
  }
})

afterAll(async () => {
  await Promise.all(
    ctx.ids.map((id: string) =>
      fetch(HELPER_SERVER + '/remove_user_data/' + id),
    ),
  )
})

it('create many entries', async () => {
  const input = gql`
    mutation create_many_entries($entries: [EntryInput!]!) {
      createManyEntries(entries: $entries) {
        id
        amount
        title
        createdAt
        updatedAt
      }
    }
  `

  const vars = {
    entries: Array.from({ length: 36 }).map((_, i) => ({
      title: `title ${i}`,
      amount: i * 1.3 + 1,
    })),
  } as MutationCreateManyEntriesArgs

  await fetchQql<
    {
      createManyEntries: Mutation['createManyEntries']
    },
    MutationCreateManyEntriesArgs
  >(ctx.user.headers, input.input, vars)

  const query = gql`
    query get_all_entries($pagination: PaginationRequest!) {
      getAllEntries(pagination: $pagination) {
        data {
          amount
          title
        }
        page
        pageSize
        total
        totalPages
      }
    }
  `

  const quick = async (page: number) => {
    return await fetchQql<
      {
        getAllEntries: Query['getAllEntries']
      },
      QueryGetAllEntriesArgs
    >(ctx.user.headers, query.input, {
      pagination: {
        page: page,
        pageSize: 10,
      },
    } satisfies QueryGetAllEntriesArgs)
  }

  const data2 = await quick(1)
  expect(data2.getAllEntries.data.length).toBe(10)
  const data3 = await quick(2)
  expect(data3.getAllEntries.data.length).toBe(10)
  const data4 = await quick(3)
  expect(data4.getAllEntries.data.length).toBe(10)
  const data5 = await quick(4)
  expect(data5.getAllEntries.data.length).toBe(6)

  expect({
    data2,
    data3,
    data4,
    data5,
  }).toMatchSnapshot()
})
