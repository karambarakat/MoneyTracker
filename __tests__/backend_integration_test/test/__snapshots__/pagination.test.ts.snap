// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create many entries 1`] = `
Object {
  "data2": Object {
    "getAllEntries": Object {
      "data": Array [
        Object {
          "amount": 46.5,
          "title": "title 35",
        },
        Object {
          "amount": 45.2,
          "title": "title 34",
        },
        Object {
          "amount": 43.9,
          "title": "title 33",
        },
        Object {
          "amount": 42.6,
          "title": "title 32",
        },
        Object {
          "amount": 41.3,
          "title": "title 31",
        },
        Object {
          "amount": 40,
          "title": "title 30",
        },
        Object {
          "amount": 38.7,
          "title": "title 29",
        },
        Object {
          "amount": 37.4,
          "title": "title 28",
        },
        Object {
          "amount": 36.1,
          "title": "title 27",
        },
        Object {
          "amount": 34.8,
          "title": "title 26",
        },
      ],
      "page": 1,
      "pageSize": 10,
      "total": 36,
      "totalPages": 4,
    },
  },
  "data3": Object {
    "getAllEntries": Object {
      "data": Array [
        Object {
          "amount": 33.5,
          "title": "title 25",
        },
        Object {
          "amount": 32.2,
          "title": "title 24",
        },
        Object {
          "amount": 30.9,
          "title": "title 23",
        },
        Object {
          "amount": 29.6,
          "title": "title 22",
        },
        Object {
          "amount": 28.3,
          "title": "title 21",
        },
        Object {
          "amount": 27,
          "title": "title 20",
        },
        Object {
          "amount": 25.7,
          "title": "title 19",
        },
        Object {
          "amount": 24.4,
          "title": "title 18",
        },
        Object {
          "amount": 23.1,
          "title": "title 17",
        },
        Object {
          "amount": 21.8,
          "title": "title 16",
        },
      ],
      "page": 2,
      "pageSize": 10,
      "total": 36,
      "totalPages": 4,
    },
  },
  "data4": Object {
    "getAllEntries": Object {
      "data": Array [
        Object {
          "amount": 20.5,
          "title": "title 15",
        },
        Object {
          "amount": 19.2,
          "title": "title 14",
        },
        Object {
          "amount": 17.9,
          "title": "title 13",
        },
        Object {
          "amount": 16.6,
          "title": "title 12",
        },
        Object {
          "amount": 15.3,
          "title": "title 11",
        },
        Object {
          "amount": 14,
          "title": "title 10",
        },
        Object {
          "amount": 12.7,
          "title": "title 9",
        },
        Object {
          "amount": 11.4,
          "title": "title 8",
        },
        Object {
          "amount": 10.1,
          "title": "title 7",
        },
        Object {
          "amount": 8.8,
          "title": "title 6",
        },
      ],
      "page": 3,
      "pageSize": 10,
      "total": 36,
      "totalPages": 4,
    },
  },
  "data5": Object {
    "getAllEntries": Object {
      "data": Array [
        Object {
          "amount": 7.5,
          "title": "title 5",
        },
        Object {
          "amount": 6.2,
          "title": "title 4",
        },
        Object {
          "amount": 4.9,
          "title": "title 3",
        },
        Object {
          "amount": 3.6,
          "title": "title 2",
        },
        Object {
          "amount": 2.3,
          "title": "title 1",
        },
        Object {
          "amount": 1,
          "title": "title 0",
        },
      ],
      "page": 4,
      "pageSize": 10,
      "total": 36,
      "totalPages": 4,
    },
  },
}
`;
